pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -312,7 +312,7 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
 	return ok && operation == v1beta1constants.ShootOperationMaintain
 }
 
-func filterForCRIName(machineImageFromCloudProfile *gardencorev1beta1.MachineImage, workerCRI *gardencorev1beta1.CRI) *gardencorev1beta1.MachineImage {
+func filterForCRI(machineImageFromCloudProfile *gardencorev1beta1.MachineImage, workerCRI *gardencorev1beta1.CRI) *gardencorev1beta1.MachineImage {
```

Comment:
I wonder what will happen in the following case:
1. No CRI is configured for worker (-> Docker for K8s <= 1.21).
2. CloudProfile contains machine images only supporting non Docker CRIs.

IIUC, this means that images which don't support Docker will be considered as valid updates. Can you double check this?

Reply:
This is very much correct. Our current situation is such that we can only start producing images without docker support once there are no longer any clusters with `cri: nil` configuration anymore.

Reply:
I see. What do you think about adding a validation for this constraint to the [admission plugin](https://github.com/gardener/gardener/blob/master/plugin/pkg/global/resourcereferencemanager/admission.go#L381)?

Reply:
I realized the behavior of `cri:nil` and `cri.name:docker` specified by the worker should be treated in the same way by the maintenance controller: update only if the image explicitly supports `docker` as a runtime.
This comes with the requirement to add this to your Cloud Profile (I've adapted the release notes above with a **breaking** note)

Reply:
Thanks for the explanation and the change. I just wonder why we shouldn't treat a CloudProfile image version without a CRI specification as it was configured with Docker, as well? This would be a backwards compatible change.

Reply:
> I just wonder why we shouldn't treat a CloudProfile image version without a CRI specification as it was configured with Docker

I guess this makes it hard to ever produce an image without docker support, which we may want to do in the future.

Reply:
I'm thinking that after your latest change, things got a little asymmetric. On the one hand you treat `cri:nil` as `cri.name:docker` for `Shoots`. But for `CloudProfiles` you don't handle `.spec.machineImages[].versions[].cri: nil` as `.spec.machineImages[].versions[].cri: [{name: containerd}]`.

I assume that today the majority of operators and users don't specify CRI (whether in `Shoot` or `CloudProfile`) at all, which means they use Docker. The issue I see with this incompatible change here is that operators will fail at a very late stage if they miss to follow the instructions (of course they should read it), i.e. an update did not happen when it was supposed to happen + there is no error message or error log which can help the operator to identify the issue. That is the reason I suggested this transition to become more operator/user friendly through one or a combination of:
- Backwards compatibility
- Automated migration, i.e. add `cri.name:docker` if `cri:nil` for `Shoots` and `CloudProfiles`
- Validation to fail early (e.g. `cri:nil` in `CloudProfiles` should not be a valid use case any more)
- Print error in log message if appropriate image could not be found

Reply:
Great points, thanks! I've added the following story plus some sub-tasks to #4110 
> As an operator, I want Gardener take care of migrating my Cloud Profile from implicit docker runtime support to explicit specification of docker, such that I don't have to do error-prone manual work myself

For the Shoot defaulting: we decided explicitly against migrating automatically to `cri.name: docker`, as we want Shoot owners to face this decision as early as possible (do I need to change my production workload, such that I can switch to `containerd` once support for `docker` is removed?). Our assumption here is that when upgrading to k8s 1.22, most people will be fine with the new default of `containerd`.